### PR TITLE
EID-1588: Refactored Metadata checker for to support eidas metadata

### PR DESCRIPTION
**Pull request by tomrosier**
Refactored tests,
Fixed crash if certificates didn't have ocsp.

### CODE REVIEW

File: bin/prometheus-metadata-exporter [hunk 1]
```diff
@@ -113,16 +113,27 @@ class OcspCheckMetric < Prometheus::Client::Gauge
   private
 
   def check_and_set(gauge, certificate_identities, ca_files)
+      return if certificate_identities.nil?
+
       pems = certificate_identities.keys
       ocsp_results = @pem_checker.check_pems(pems, ca_files)
       certificate_identities.map do |pem, identities|
         identities.map do |identity|
+          puts(identity)
```
Comment (instantiator) [inline bin/prometheus-metadata-exporter]: Is this debug logging, and should it stay in?

Comment (tomrosier) [reply bin/prometheus-metadata-exporter]: Oops yes il remove that.

File: bin/prometheus-metadata-exporter [hunk 2]
```diff
@@ -113,16 +113,27 @@ class OcspCheckMetric < Prometheus::Client::Gauge
...
           cert = @certificate_factory.from_inline_pem(pem)
+
+          ocsp_result = 1
+          if ocsp_results[pem].unknown?
+            ocsp_result = -1
```
Comment (philandstuff) [inline bin/prometheus-metadata-exporter]: :thinking: to fill in

Comment (philandstuff) [inline bin/prometheus-metadata-exporter]: This isn't a very prometheus-style way of encoding information.  Right now we have a metric `verify_metadata_certificate_ocsp_success` which is `1` if ocsp says the certificate is valid and `0` if ocsp says the certificate is invalid.  Adding `-1` as a third state to the metric feels wrong.

Alternatives I would suggest:
1. don't emit a metric at all if OCSP information isn't present (ie skip the `gauge.set()` call below)
2. emit a separate metric which encodes whether OCSP information is present `verify_metadata_certificate_ocsp_known` where `1` means "known" and `0` means "unknown"

I'd probably go with 1 unless we have a need for this information.

File: features/metadata_checks_are_reported_in_a_prometheus_compatible_metrics_endpoint.feature [hunk 3]
```diff
@@ -20,56 +20,56 @@ Feature: metadata checks are reported in a prometheus compatible metrics endpoin
...
-    verify_metadata_expiry{metadata="http://localhost:53110"} 1434720100000.0
-    # TYPE verify_metadata_certificate_ocsp_success gauge
-    # HELP verify_metadata_certificate_ocsp_success If a cert chain validation and OCSP check of the given X.509 SAML certificate is good \(1\) or bad \(0\)
-    verify_metadata_certificate_ocsp_success{entity_id="foo",use="encryption",key_name="foo_key_1",serial="2",subject="/DC=org/DC=TEST/CN=GENERATED TEST CERTIFICATE"} 1.0
-    verify_metadata_certificate_ocsp_success{entity_id="foo",use="encryption",key_name="foo_key_2",serial="3",subject="/DC=org/DC=TEST/CN=GENERATED TEST CERTIFICATE"} 1.0
-    verify_metadata_certificate_ocsp_success{entity_id="bar",use="encryption",key_name="bar_key_1",serial="4",subject="/DC=org/DC=TEST/CN=GENERATED TEST CERTIFICATE"} 1.0
-    verify_metadata_certificate_ocsp_success{entity_id="metadata_signature",use="",key_name="certificate",serial="2",subject="/DC=org/DC=TEST/CN=SIGNED TEST CERTIFICATE"} 1.0
+verify_metadata_certificate_expiry{entity_id="foo",use="encryption",serial="2",subject="/DC=org/DC=TEST/CN=GENERATED TEST CERTIFICATE"} (\d+).0
```
Comment (philandstuff) [inline features/metadata_checks_are_reported_in_a_prometheus_compatible_metrics_endpoint.feature]: These should be indented to the same level as the surrounding `"""` ([source](https://cucumber.io/docs/gherkin/reference/) which says "The indentation inside the triple quotes, however, is significant. Each line of the Doc String will be dedented according to the opening `"""`.")

File: lib/metadata/ocsp/client.rb [hunk 4]
```diff
@@ -16,13 +16,23 @@ def check(certificates, issuer, store)
...
         }
         ocsp_uri = find_ocsp_uri(certificates.first)
-        ocsp_request = build_ocsp_request_body(cert_ids.values)
-        ocsp_response = make_ocsp_request(ocsp_uri, ocsp_request)
-        cert_id_statuses = check_response(ocsp_request, ocsp_response, cert_ids.values, store)
-        certificates.inject({}) { |hash, certificate|
-          hash[certificate] = cert_id_statuses[cert_ids[certificate]]
-          hash
-        }
+        unless ocsp_uri.nil?
```
Comment (philandstuff) [inline lib/metadata/ocsp/client.rb]: stylistically, `unless foo.nil? ... else` has three negatives in it which makes it hard to understand.  I think I'd prefer at least doing `if foo.nil? .. else` and swapping the bodies round but there may be other ways to express this.

File: lib/metadata/ocsp/client.rb [hunk 5]
```diff
@@ -69,7 +82,7 @@ def check_response(ocsp_request, ocsp_response, cert_ids, store)
         basic_response = ocsp_response.basic
         fail CheckerError, "could not verify response against issuer certificates" unless basic_response.verify([], store)
         statuses = basic_response.status
-        results = statuses.inject({}) do |hash,status|
+          results = statuses.inject({}) do |hash,status|
```
Comment (philandstuff) [inline lib/metadata/ocsp/client.rb]:
```suggestion
        results = statuses.inject({}) do |hash,status|
```
This looks like a spurious change

**Pull request merged**